data_structures/avltree.cpp

```diff
@@ -1,198 +1,426 @@
 /**
  * \file
- * \brief A simple tree implementation using nodes
- *
- * \todo update code to use C++ STL library features and OO structure
- * \warning This program is a poor implementation and does not utilize any of
- * the C++ STL features.
+ * \brief An implementation of AVL tree.
  */
 #include <algorithm>  /// for std::max
+#include <cassert>    /// for assert
 #include <iostream>   /// for std::cout
+#include <memory>     /// for std::unique_ptr
 #include <queue>      /// for std::queue
+#include <sstream>    /// for std::ostringstream
+#include <vector>     /// for std::vector
 
-using node = struct node {
-    int data;
-    int height;
-    struct node *left;
-    struct node *right;
-};
+class AVLTree {
```

Comment:
classes should be documented check [contribution guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)